src/auto-evo/EvolutionaryTree.cs

```diff
@@ -0,0 +1,224 @@
+            var lineEnd = new Vector2(GENERATION_SEPARATION * latestGeneration + latestNode.RectSize.x, lineStart.y);
+            DrawLine(lineStart, lineEnd);
+            DrawString(latoSmall, lineEnd + new Vector2(SPECIES_NAME_OFFSET, 0), speciesNames[latestNode.SpeciesID]);
+        }
+
+        // Draw extinct species name
+        foreach (var extinctedSpecies in latestNodes.Values.Where(n => n.LastGeneration))
+        {
+            DrawString(latoSmall, new Vector2(
+                extinctedSpecies.RectPosition.x + extinctedSpecies.RectSize.x + SPECIES_NAME_OFFSET,
+                extinctedSpecies.Center.y), speciesNames[extinctedSpecies.SpeciesID], Colors.DarkRed);
+        }
+    }
+
+    public void UpdateEvolutionaryTreeWithRunResults(RunResults results, int generation)
+    {
+        foreach (var speciesResultPair in results.OrderBy(r => r.Value.Species.ID))
+        {
+            var species = speciesResultPair.Key;
+            var result = speciesResultPair.Value;
+
+            if (result.Species.Population <= 0)
+            {
+                if (result.SplitFrom == null)
+                {
+                    SetupTreeNode((Species)species.Clone(),
+                        nodes.FindLast(n => n.SpeciesID == species.ID), generation - 1, true);
+                }
+            }
+            else if (result.SplitFrom != null)
+            {
+                SetupTreeNode((Species)species.Clone(),
+                    nodes.FindLast(n => n.SpeciesID == result.SplitFrom.ID), generation);
+
+                speciesOrigin.Add(species.ID, (result.SplitFrom.ID, generation));
+                speciesNames.Add(species.ID, species.FormattedName);
+            }
+            else if (result.MutatedProperties != null)
+            {
+                SetupTreeNode((Species)result.MutatedProperties.Clone(),
+                    nodes.FindLast(n => n.SpeciesID == species.ID), generation);
+            }
+
+            if (species.ID > maxSpeciesId)
+                maxSpeciesId = species.ID;
+        }
+
+        latestGeneration = generation;
+
+        BuildTree();
+        AdjustSize();
+    }
+
+    private void AdjustSize()
+    {
+        RectMinSize = new Vector2(GENERATION_SEPARATION * latestGeneration + 200,
+            SPECIES_SEPARATION * maxSpeciesId + 200);
+    }
+
+    private void SetupTreeNode(Species species, EvolutionaryTreeNode? parent, int generation,
+        bool isLastGeneration = false)
+    {
+        var node = treeNodeScene.Instance<EvolutionaryTreeNode>();
+        node.Generation = generation;
+        node.SpeciesID = species.ID;
+        node.LastGeneration = false;
+        node.ParentNode = parent;
+        node.RectPosition = new Vector2(generation * GENERATION_SEPARATION, TIMELINE_HEIGHT);
+        node.LastGeneration = isLastGeneration;
+        node.Group = nodesGroup;
+        node.Connect("button_down", this, nameof(OnTreeNodeSelected), new Array { node });
```

Comment:
Shouldn't this be on the clicked event as the tree node is a button so it will emit the button event whenever things that constitute clicking a button happen?